src/pybind/mgr/cephadm/upgrade.py

```diff
@@ -349,6 +349,46 @@ def _prepare_for_mds_upgrade(
 
         return continue_upgrade
 
+    def _post_upgrade_redeploys(self):
+        # type: () -> None
+        daemons = self.mgr.cache.get_daemons()
+
+        # redeploy all daemons who are not redployed during upgrade
+        # and mgrs which could have been deployed by a mgr running old code
+        for d in daemons:
+            assert d.daemon_type is not None
+            assert d.daemon_id is not None
+            assert d.hostname is not None
+
+            if d.daemon_type in CEPH_UPGRADE_ORDER and d.daemon_type != 'mgr':
+                continue
+
+            # current active mgr cannot redeploy itself. Schedule it for redeploy
+            # later then continue redeploying other daemons.
+            if self.mgr.daemon_is_self(d.daemon_type, d.daemon_id):
+                self.mgr.cache.schedule_daemon_action(d.hostname, d.name(), 'redeploy')
+                continue
+
+            if not self._wait_for_ok_to_stop(d):
+                return
```

Comment:
If you return here, `_post_upgrade_redeploys` returns and `_do_upgrade` continues with finishing the upgrade process. If you return here, we're not going to retry redeploying the daemons.